config/config/src/lib.rs

```diff
@@ -219,45 +222,52 @@ impl<C: Config> Default for ConfigBuilder<C> {
     }
 }
 
-fn merge(first: Value, second: Value) -> (Value, Option<ConfigError>) {
-    let values = (first, second);
-    let (Value::Map(mut first), Value::Map(mut second)) = values else {
-        return (values.0, None);
-    };
-
-    let mut merged = BTreeMap::new();
-    // Get all unique keys from both maps
-    let keys = first
-        .keys()
-        .chain(second.keys())
-        .cloned()
-        .collect::<HashSet<_>>();
-
-    let mut error: Option<ConfigError> = None;
-
-    for key in keys {
-        let first = first.remove(&key);
-        let second = second.remove(&key);
-
-        let (value, new_error) = match (first, second) {
-            (Some(first), Some(second)) => merge(first, second),
-            (Some(first), None) => (first, None),
-            (None, Some(second)) => (second, None),
-            (None, None) => unreachable!(),
-        };
-
-        error = merge_error_opts(error, new_error);
-        merged.insert(key, value);
+fn merge(first: Value, second: Value) -> Value {
+    match (first, second) {
+        (Value::Map(mut first), Value::Map(mut second)) => {
+            let mut merged = BTreeMap::new();
+            // Get all unique keys from both maps
+            let keys = first
+                .keys()
+                .chain(second.keys())
+                .cloned()
+                .collect::<HashSet<_>>();
+        
+            for key in keys {
+                let first = first.remove(&key);
+                let second = second.remove(&key);
+        
+                let value = match (first, second) {
+                    (Some(first), Some(second)) => merge(first, second),
+                    (Some(first), None) => first,
+                    (None, Some(second)) => second,
+                    (None, None) => unreachable!(),
+                };
+        
+                merged.insert(key, value);
+            }
+        
+            Value::Map(merged)
+        },
```

Comment:
```suggestion
         (Value::Map(first), Value::Map(second)) => {
            let mut merged = BTreeMap::new();
            let mut first = first.into_iter();
            let mut second = second.into_iter();

            loop {
                match (first.next(), second.next()) {
                    (None, None) => break,
                    (Some((k, v)), None) => {
                        merged.insert(k, v);
                    }
                    (None, Some((k, v))) => {
                        merged.insert(k, v);
                    }
                    (Some((k1, v1)), Some((k2, v2))) => {
                        if k1 == k2 {
                            merged.insert(k1, merge(v1, v2));
                        } else {
                            merged.insert(k1, v1);
                            merged.insert(k2, v2);
                        }
                    }
                }
            }

            Value::Map(merged)
        }
```

Reply:
you could use [zip_longest](https://docs.rs/itertools/0.8.0/itertools/trait.Itertools.html#method.zip_longest) if you want to be more concise